zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkConnection.java

```diff
@@ -102,6 +112,7 @@ protected void reconnect(Watcher watcher) throws InterruptedException {
         LOG.debug("Creating new ZookKeeper instance to reconnect to " + _servers + ".");
         _zk = new ZooKeeper(_servers, _sessionTimeOut, watcher);
         prevZk.close();
+        _getChildrenMethod = null;
```

Comment:
Not sure if this will cause some race condition issues?

For example, the reconnect() cleans up the method here. And in the other application thread, the operation is about to use this object (with the same reference), then it will be an NPE.

Reply:
Theoretically it could happen, though the possibility is almost 0. It’s like
```
if (a == null) { a = xxx; }
a.call();
```
Between these two lines.
NPE does not cause correctness issue, the operation just fails. NPE could also happen now: like zkclient is closed, another thread is still trying the zk operation, it could also get NPE, as zk = null in close()

There are a few options to avoid it:

1. lock the method reference. Obviously we don’t want it as it’s not nice. Would ask why only getChildren is locked but not other APIs. Kind of weird..
2. Instead of setting method = null in reconnect, we do lookup() and assign a value to it, so null is avoided. This is something I considered before when comparing it with setting to null (I felt setting to null might be risky not a good practice at least..)
3. Just don’t reset method in reconnect. The purpose of resetting is for client has pagination zk version first, and then zk server upgrades, so client can reset and detect server supports it then client can start to use pagination. But it could not cover all cases, like if servers can upgrade really fast without expiring client session, client can still not detect server’s upgrade. So I think the right way is, if we want to onboard pagination, always expect server upgrades first then client. Vice versa, pagination is not leveraged in helix client.

2 is doable, but I prefer 3, because we always follow the rule: "if we want to onboard pagination, always expect server upgrades first then client".  2 is not transparent or consistent: sometimes it can detect and sometimes not.

What do you think?